kubernetes/config/kube_config.py

```diff
@@ -22,151 +22,257 @@
+            self._client_configuration = client_configuration
+        else:
+            self._client_configuration = configuration
+
+    def set_active_context(self, context_name=None):
+        if context_name is None:
+            context_name = self._config['current-context']
+        self._current_context = self._config['contexts'].get_with_name(
+            context_name)
+        if self._current_context['context'].safe_get('user'):
+            self._user = self._config['users'].get_with_name(
+                self._current_context['context']['user'])['user']
+        else:
+            self._user = None
+        self._cluster = self._config['clusters'].get_with_name(
+            self._current_context['context']['cluster'])['cluster']
+
+    def _load_authentication(self):
+        """Read authentication from kube-config user section if exists.
+
+        This function goes through various authetication methods in user
+        section of kubeconfig and stops if it founds a valid authentication
+        method. The order of authentication methods is:
+
+            1. GCP auth-provider
+            2. token_data
+            3. token field (point to a token file)
+            4. username/password
+        """
+        if not self._user:
+            return
+        if self._load_gcp_token():
+            return
+        if self._load_user_token():
+            return
+        self._load_userpass_token()
+
+    def _load_gcp_token(self):
+        if 'auth-provider' not in self._user:
+            return
+        if 'name' not in self._user['auth-provider']:
+            return
+        if self._user['auth-provider']['name'] != 'gcp':
+            return
+        # Ignore configs in auth-provider and rely on GoogleCredentials
+        # caching and refresh mechanism.
+        # TODO: support gcp command based token ("cmd-path" config).
+        self.token = self._get_google_credentials()
+        return self.token
+
+    def _load_user_token(self):
+        token = FileOrData(self._user, 'tokenFile', 'token').data
+        if token:
+            self.token = token
+            return True
+
+    def _load_userpass_token(self):
+        if 'username' in self._user and 'password' in self._user:
+            self.token = urllib3.util.make_headers(
+                basic_auth=(self._user['username'] + ':' +
+                            self._user['password'])).get('authorization')
+            return True
+
+    def _load_cluster_info(self):
+        if 'server' in self._cluster:
+            self.host = self._cluster['server']
+            if self.host.startswith("https"):
+                self.ssl_ca_cert = FileOrData(
+                    self._cluster, 'certificate-authority').file
+                self.cert_file = FileOrData(
+                    self._user, 'client-certificate').file
+                self.key_file = FileOrData(self._user, 'client-key').file
+
+    def _set_config(self):
+        if hasattr(self, 'token'):
```

Comment:
Using self.__dict__ would work in this case because 'token' is an instance attribute.  I'm not sure I'd want to get into the habit of that approach, though, since it wouldn't play nice with class attributes, inherited or otherwise:

```python
class foo(object):
 x = 1

class bar(foo):
 y = 1

>>> obj = bar()
>>> obj.__dict__
{}
>>> obj.x
1
>>> obj.x = 1
1
>>> obj.__dict__
{'x': 1}
```

Using either ``getattr()`` or catching AttributeError will work regardless of where something is defined, and I think that having fewer ways to do things makes for code that is easier to maintain.